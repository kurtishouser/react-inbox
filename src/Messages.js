import React, { Component } from 'react';
import Message from './Message.js';

class Messages extends Component {

  render() {
    return (
      <div>
        {this.props.messages.map((message) => {
          return <Message
                    key={message.id}
                    id={message.id}
                    read={message.read}
                    starred={message.starred}
                    selected={message.selected}
                    labels={message.labels}
                    subject={message.subject}
                    updateSelectedStatus={this.props.updateSelectedStatus}
                    updateStarredStatus={this.props.updateStarredStatus}
                    deleteMessages={this.props.deleteMessages}
                    toggleProperty={this.props.toggleProperty}/>
        })}
      </div>
    )
  }
}

export default Messages;
