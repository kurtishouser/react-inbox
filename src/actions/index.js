export const MESSAGES_PENDING = 'MESSAGES_PENDING';
export const MESSAGES_RECEIVED = 'MESSAGES_RECEIVED';
export function fetchMessages() {
  return async (dispatch, getState, { Api }) => {

    if (getState().messages.ids.length) {
      return
    }

    dispatch({
      type: MESSAGES_PENDING
    });

    const json = await Api.fetchMessages();

    return dispatch({
      type: MESSAGES_RECEIVED,
      messages: json._embedded.messages
    });
  }
}

export const SEND_MESSAGE = 'SEND_MESSAGE';
export function sendMessage(message) {
  return async (dispatch, getState, { Api }) => {

    let body = message;

    const response = await Api.sendMessage(body);

    return dispatch({
      type: SEND_MESSAGE,
      response
    });
  }
}

export const FETCH_MESSAGE_BODY = 'FETCH_MESSAGE_BODY';
export function fetchMessageBody(id) {
  return async (dispatch, getState, { Api }) => {

  if (getState().messages.messagesById[id].body.length > 0) {
    // update read status just in case
    await Api.patchRequest({
      'messageIds': [id],
        'command': 'read',
        'read': true,
    });

    return;
  }

    const response = await Api.fetchMessageBody(id);

    // update read status
    await Api.patchRequest({
      'messageIds': [id],
        'command': 'read',
        'read': true,
    });

    return dispatch({
      type: FETCH_MESSAGE_BODY,
      id: response.id,
      messageBody: response.body
    })
  }
}

export const TOGGLE_SELECTED = 'TOGGLE_SELECTED';
export function toggleSelected(id) {
  return {
      type: TOGGLE_SELECTED,
      id
  }
}

export const TOGGLE_STARRED = 'TOGGLE_STARRED';
export function toggleStarred(id) {
  return async (dispatch, getState, { Api }) => {

    let body = {
        'messageIds': [id],
        'command': 'star',
        'star': !getState().messages.messagesById[id].starred,
    }

    const response = await Api.patchRequest(body);

    if (response === 200) {
      return dispatch({
        type: TOGGLE_STARRED,
        id
      });
    }
  }
}

export const TOGGLE_SELECT_ALL ='TOGGLE_SELECT_ALL';
export function toggleSelectAll(status) {
  return {
    type: TOGGLE_SELECT_ALL,
    status
  }
}

export const UPDATE_READ_STATUS = 'UPDATE_READ_STATUS';
export function updateReadStatus(status) {
  return async (dispatch, getState, { Api }) => {

    let messageIds = getState().messages.ids.filter(id => (
      getState().messages.messagesById[id].selected
      // && getState().messages.messagesById[id].read !== status
    ));

    // only make the API call and dispatch action when there are IDs
    if (messageIds.length) {
      let body = {
        'messageIds': messageIds,
          'command': 'read',
          'read': status,
        }

      const response = await Api.patchRequest(body);

      if (response === 200) {
        return dispatch({
          type: UPDATE_READ_STATUS,
          status,
          messageIds // optimizes reducer
        });
      }
    }
  }
}

export const ADD_LABEL = 'ADD_LABEL';
export function addLabel(label) {
  return async (dispatch, getState, { Api }) => {

    let messageIds = getState().messages.ids.filter(id => (
      getState().messages.messagesById[id].selected &&
      !getState().messages.messagesById[id].labels.includes(label)));

    // only make the API call and dispatch action when there are IDs
    if (messageIds.length) {
      let body = {
        'messageIds': messageIds,
        'command': 'addLabel',
        'label': label,
      }

      const response = await Api.patchRequest(body)

      if (response === 200) {
        return dispatch({
          type: ADD_LABEL,
          label,
          messageIds // optimizes reducer
        });
      }
    }
  }
}

export const REMOVE_LABEL = 'REMOVE_LABEL';
export function removeLabel(label) {
  return async (dispatch, getState, { Api }) => {

    let messageIds = getState().messages.ids.filter(id => (
      getState().messages.messagesById[id].selected &&
      getState().messages.messagesById[id].labels.includes(label)));

    // only make the API call and dispatch action when there are IDs
    if (messageIds.length) {
      let body = {
        'messageIds': messageIds,
        'command': 'removeLabel',
        'label': label,
      }

      const response = await Api.patchRequest(body)

      if (response === 200) {
        return dispatch({
          type: REMOVE_LABEL,
          label,
          messageIds // optimizes reducer
        });
      }
    }
  }
}

export const DELETE_MESSAGES= 'DELETE_MESSAGES';
export function deleteMessages() {
  return async (dispatch, getState, { Api }) => {

    let messageIds = getState().messages.ids.filter(id => (
      getState().messages.messagesById[id].selected));

    if (messageIds.length) {
      let body = {
        'messageIds': messageIds,
        'command': 'delete',
      }

      const response = await Api.patchRequest(body)

      if (response === 200) {
        return dispatch({
          type: DELETE_MESSAGES,
        });
      }
    }
  }
}
