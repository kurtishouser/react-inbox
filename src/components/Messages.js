import React, { Component } from 'react';
import Message from './Message.js';
import { connect } from 'react-redux';
import { fetchMessages } from '../actions';
import { bindActionCreators } from 'redux';

class Messages extends Component {

  componentDidMount() {
    this.props.fetchMessages();
  }

  render() {
    const { loading, messageIds } = this.props;
    return !loading ?
      (
        <div>
          { messageIds.map((messageId) => <Message key={messageId} messageId={messageId} />) }
        </div>
      ) :
      (<div>Loading...</div>)
  }
}

const mapStateToProps = state => {
  const loading = state.messages.messagesLoading;
  const messageIds = state.messages.ids;
  return {
    loading,
    messageIds
  }
}

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchMessages
}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Messages);
