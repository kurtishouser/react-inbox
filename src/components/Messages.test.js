import React from 'react'
import { shallow, render, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Messages, mapStateToProps } from './Messages'
import { fetchMessages, MESSAGES_RECEIVED } from '../actions'

describe('Messages', () => {

  it('should render without any props defined', () => {
    const component = shallow(
      <Messages messageIds={[]}/>
    );
    expect(toJson(component)).toMatchSnapshot();
  })

  it('Calls fetchMessages on load', () => {
    let fetchMessages = jest.fn()
    const component = mount(
      <Messages messageIds={[]}/>
    )
    expect(fetchMessages).toHaveBeenCalled();
  });

});
