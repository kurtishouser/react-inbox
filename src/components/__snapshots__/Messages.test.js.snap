// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages should render without any props defined 1`] = `<div />`;
